text/0119-standard-system-metrics.md

```diff
@@ -0,0 +1,146 @@
+- **time** - an instrument that measures passage of time should be called `entity.time`. For example, `system.cpu.time` with varying values of label `state` for idle, user, etc.
+- **io** - an instrument that measures bidirectional data flow should be called `entity.io` and have labels for direction. For example, `system.net.io`.
+- Other metrics that do not fit the above descriptions may be named more freely. For example, `system.swap.page_faults` and `system.net.packets`. Units do not need to be specified in the names since they are included during instrument creation, but can be added if there is ambiguity.
+
+## Internal details
+
+The following standard metric names should be used in libraries instrumenting system/runtime metrics (here is a [spreadsheet](https://docs.google.com/spreadsheets/d/1r50cC9ass0A8SZIg2ZpLdvZf6HmQJsUSXFOu-rl4yaY/edit#gid=973941697) of the tables below).
+
+### Standard System Metrics - `system.`
+
+---
+
+#### `system.cpu.`
+
+**Description:** System level processor metrics.
+|Name                  |Units  |Instrument       |Label Key|Label Values                       |
+|----------------------|-------|-----------------|---------|-----------------------------------|
+|system.cpu.time       |seconds|SumObserver      |state    |idle, user, system, interrupt, etc.|
+|                      |       |                 |cpu      |1 - #cores                         |
+|system.cpu.utilization|%      |UpDownSumObserver|state    |idle, user, system, interrupt, etc.|
+|                      |       |                 |cpu      |1 - #cores                         |
+
+#### `system.memory.`
+
+**Description:** System level memory metrics.
+|Name                     |Units|Instrument       |Label Key|Label Values            |
+|-------------------------|-----|-----------------|---------|------------------------|
+|system.memory.usage      |bytes|UpDownSumObserver|state    |used, free, cached, etc.|
+|system.memory.utilization|%    |UpDownSumObserver|state    |used, free, cached, etc.|
+
+#### `system.swap.`
+
+**Description:** System level swap/paging metrics.
+|Name                    |Units|Instrument       |Label Key|Label Values|
+|------------------------|-----|-----------------|---------|------------|
+|system.swap.usage       |pages|UpDownSumObserver|state    |used, free  |
+|system.swap.utilization |%    |UpDownSumObserver|state    |used, free  |
+|system.swap.page\_faults|1    |SumObserver      |type     |major, minor|
+|system.swap.paging\_ops |1    |SumObserver      |type     |major, minor|
+|                        |     |                 |direction|in, out     |
+
+#### `system.disk.`
+
+**Description:** System level disk performance metrics.
+|Name                        |Units|Instrument |Label Key|Label Values|
+|----------------------------|-----|-----------|---------|------------|
+|system.disk.io<!--notlink-->|bytes|SumObserver|device   |(identifier)|
```

Comment:
I wonder how often a user will meaningfully aggregate this metric and combine both the `read` and `write` values. It's a question of whether this is a meaningful enough aggregation that we should use a single metric. I'd lean in the other direction, myself. Could we use one metric for read and one for write? (The same for all the bi-directional metrics in this document.)

Reply:
I am not sure on this one, could something like ratio of reads/writes be useful? Just want to copy [that prometheus recommendation](https://prometheus.io/docs/practices/naming/#metric-names) we keep going back to here:
> As a rule of thumb, either the sum() or the avg() over all dimensions of a given metric should be meaningful (though not necessarily useful). If it is not meaningful, split the data up into multiple metrics. For example, having the capacity of various queues in one metric is good, while mixing the capacity of a queue with the current number of elements in the queue is not.

I think disk IO falls into the category of meaningful but not necessarily useful? The sum would be the total number of bytes

Reply:
I agree with your logic. Let's keep it a single metric.